File: ArchiveProcessor/ArchiveCaptureiOS/Sources/ArchiveCaptureiOS/Net/RelayObjectFormat.swift
import Foundation
import CryptoKit

/// iOS mirror of the Mac's `RelayObjectFormat` — MUST produce byte-identical output (canonical JSON, escape
/// table, fingerprint) so the Mac receiver reads what this writes. Guarded by the golden byte-check against
/// `SPEC/relay-golden/`. Keep in lockstep with `ArchiveProcessor/.../Net/RelayObjectFormat.swift` and the
/// Android `net/RelayObjectFormat.kt`. Governing spec: `LIVE_CAPTURE_FILERELAY_SPEC.md` (v2 amendments bind).
enum RelayObjectFormat {

    // MARK: - Object names
    static func jpegName(group: String, seq: Int) -> String { "\(group)__\(seq).jpg" }
    static func sidecarName(group: String, seq: Int) -> String { "\(group)__\(seq).json" }
    static func receiptName(group: String, seq: Int) -> String { "\(group)__\(seq).receipt.json" }
    static func segmentName(group: String) -> String { "\(group).segment.json" }
    static let sessionCompleteName = "_session.complete.json"
    static let epochMarkerName = "_epoch.json"

    // MARK: - Canonical JSON (all-string values, sorted keys, nil-omitted, fixed escaping)
    static func canonicalJSON(_ map: [String: String?]) -> Data {
        let pairs = map.compactMap { (k, v) -> (String, String)? in v.map { (k, $0) } }
                       .sorted { $0.0 < $1.0 }
        var s = "{"
        for (i, pair) in pairs.enumerated() {
            if i > 0 { s += "," }
            s += "\"" + escape(pair.0) + "\":\"" + escape(pair.1) + "\""
        }
        s += "}"
        return Data(s.utf8)
    }

    private static func escape(_ s: String) -> String {
        var out = ""
        out.reserveCapacity(s.count)
        for scalar in s.unicodeScalars {
            switch scalar {
            case "\"": out += "\\\""
            case "\\": out += "\\\\"
            case "\u{08}": out += "\\b"
            case "\u{09}": out += "\\t"
            case "\u{0A}": out += "\\n"
            case "\u{0C}": out += "\\f"
            case "\u{0D}": out += "\\r"
            default:
                if scalar.value < 0x20 { out += String(format: "\\u%04x", scalar.value) }
                else { out.unicodeScalars.append(scalar) }
            }
        }
        return out
    }

    // MARK: - Metadata fingerprint (A1) — identical function+inputs on Mac & phone → same fp
    static func fingerprint(type: String, priority: String?, year: String?, month: String?, replaces: String?) -> String {
        let m: [String: String?] = ["type": type, "priority": priority, "year": year, "month": month, "replaces": replaces]
        let digest = SHA256.hash(data: canonicalJSON(m))
        return digest.prefix(8).map { String(format: "%02x", $0) }.joined()
    }

    // MARK: - Encode (what the phone writes)
    static func encodeSidecar(token: String, epoch: String, group: String, seq: Int, type: String,
                              priority: String?, year: String?, month: String?, replaces: String?,
                              device: String? = nil) -> Data {
        let fp = fingerprint(type: type, priority: priority, year: year, month: month, replaces: replaces)
        return canonicalJSON(["kind": "photo", "token": token, "epoch": epoch, "group": group,
                              "seq": String(seq), "type": type, "priority": priority, "year": year,
                              "month": month, "replaces": replaces, "device": device, "fp": fp])
    }

    static func encodeSegment(token: String, epoch: String, group: String,
                              priority: String?, year: String?, month: String?, seqs: String?) -> Data {
        canonicalJSON(["kind": "segment-complete", "token": token, "epoch": epoch, "group": group,
                       "priority": priority, "year": year, "month": month, "seqs": seqs])
    }

    static func encodeSessionComplete(token: String, epoch: String) -> Data {
        canonicalJSON(["kind": "session-complete", "token": token, "epoch": epoch])
    }

    // MARK: - Parse (lenient — number-or-string tolerated). The phone reads the receipt + the epoch marker.
    static func parse(_ data: Data) -> [String: String]? {
        guard let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { return nil }
        var out: [String: String] = [:]
        for (k, v) in obj {
            if let s = v as? String { out[k] = s }
            else if let n = v as? NSNumber { out[k] = n.stringValue }
        }
        return out
    }
}

File: ArchiveProcessor/ArchiveCapture/app/src/main/java/com/archiveprocessor/capture/net/RelayObjectFormat.kt
package com.archiveprocessor.capture.net

import java.security.MessageDigest

/**
 * Pure-Kotlin (no Android API) mirror of the Mac/iOS `RelayObjectFormat`. MUST produce byte-identical
 * canonical JSON so the Mac receiver reads what this writes — guarded by the golden byte-check against
 * `SPEC/relay-golden/`. Keep in lockstep with `ArchiveProcessor/.../Net/RelayObjectFormat.swift` and
 * `ArchiveCaptureiOS/.../Net/RelayObjectFormat.swift`. Spec: `LIVE_CAPTURE_FILERELAY_SPEC.md` (v2 binds).
 *
 * No `org.json` here (keeps it JVM-unit-testable); the transport does its own receipt/epoch parsing.
 */
object RelayObjectFormat {
    fun jpegName(group: String, seq: Int) = "${group}__${seq}.jpg"
    fun sidecarName(group: String, seq: Int) = "${group}__${seq}.json"
    fun receiptName(group: String, seq: Int) = "${group}__${seq}.receipt.json"
    fun segmentName(group: String) = "${group}.segment.json"
    const val SESSION_COMPLETE_NAME = "_session.complete.json"
    const val EPOCH_MARKER_NAME = "_epoch.json"

    /** Canonical JSON: nil-omitted, keys ascending (fixed lowercase-ASCII keys → same sort as Swift),
     *  fixed escape table, UTF-8 encoded once at the end (astral scalars stay identical UTF-8 sequences). */
    fun canonicalJson(map: Map<String, String?>): ByteArray {
        val pairs = map.entries.filter { it.value != null }.sortedBy { it.key }
        val sb = StringBuilder("{")
        for ((i, e) in pairs.withIndex()) {
            if (i > 0) sb.append(',')
            sb.append('"').append(escape(e.key)).append("\":\"").append(escape(e.value!!)).append('"')
        }
        sb.append('}')
        return sb.toString().toByteArray(Charsets.UTF_8)
    }

    /** Matches the Swift escape table exactly: only `"` `\` and C0 controls are escaped (short forms for
     *  BS/TAB/LF/FF/CR, `\u00xx` lowercase for the rest); `/` and all non-ASCII emitted verbatim. */
    private fun escape(s: String): String {
        val sb = StringBuilder(s.length)
        for (c in s) {
            when {
                c == '"' -> sb.append("\\\"")
                c == '\\' -> sb.append("\\\\")
                c.code == 0x08 -> sb.append("\\b")
                c.code == 0x09 -> sb.append("\\t")
                c.code == 0x0A -> sb.append("\\n")
                c.code == 0x0C -> sb.append("\\f")
                c.code == 0x0D -> sb.append("\\r")
                c.code < 0x20 -> sb.append("\\u%04x".format(c.code))
                else -> sb.append(c)
            }
        }
        return sb.toString()
    }

    /** fp = SHA-256(canonicalJSON of the ingest-relevant metadata) → first 16 hex. Same fn+inputs as the
     *  Swift side → identical fp. Excludes `device` (a device-name change must not force a re-ingest, A1). */
    fun fingerprint(type: String, priority: String?, year: String?, month: String?, replaces: String?): String {
        val m = mapOf("type" to type, "priority" to priority, "year" to year, "month" to month, "replaces" to replaces)
        val digest = MessageDigest.getInstance("SHA-256").digest(canonicalJson(m))
        return digest.copyOfRange(0, 8).joinToString("") { "%02x".format(it.toInt() and 0xFF) }
    }

    fun encodeSidecar(token: String, epoch: String, group: String, seq: Int, type: String,
                      priority: String?, year: String?, month: String?, replaces: String?, device: String? = null): ByteArray {
        val fp = fingerprint(type, priority, year, month, replaces)
        return canonicalJson(mapOf("kind" to "photo", "token" to token, "epoch" to epoch, "group" to group,
            "seq" to seq.toString(), "type" to type, "priority" to priority, "year" to year, "month" to month,
            "replaces" to replaces, "device" to device, "fp" to fp))
    }

    fun encodeSegment(token: String, epoch: String, group: String,
                      priority: String?, year: String?, month: String?, seqs: String?): ByteArray =
        canonicalJson(mapOf("kind" to "segment-complete", "token" to token, "epoch" to epoch, "group" to group,
            "priority" to priority, "year" to year, "month" to month, "seqs" to seqs))

    fun encodeSessionComplete(token: String, epoch: String): ByteArray =
        canonicalJson(mapOf("kind" to "session-complete", "token" to token, "epoch" to epoch))
}
